README.md
SignUp
======

General
-------

This is the fourth version of the SignUp Service.

 - It is written in Scala and based on the Play Framework.
 - Anorm is used for SQL Database access*
 - Heroku is used for deployment

*) H2 when running locally, PostgreSQL on Heroku


### Play Framework ###

 - Play 2.1.0 - http://www.playframework.org/
 - Scala 2.10.0 (part of Play) - http://www.scala-lang.org/
 - Anorm (part of Play) - http://www.playframework.com/documentation/2.1.0/ScalaAnorm

### Presentation ###

 - Twitter Bootstrap 2.2.1 - http://twitter.github.com/bootstrap
 - jQuery 1.9.0 (used by Twitter Bootstrap) - http://jquery.com
 - bootstrap-wysihtml5 0.0.2 - http://jhollingworth.github.com/bootstrap-wysihtml5
 - wysihtml5 0.3.0 (used by bootstrap-wysihtml5) - http://github.com/xing/wysihtml5

### Services ###
 - Play2.x module for Authentication and Authorization 0.9 - http://github.com/t2v/play20-auth
 - Emailer Plugin 2.1.0 - http://github.com/typesafehub/play-plugins/tree/master/mailer


License, credits and stuff
--------------------------

The Twitter Bootstrap LESS files that are included in the source code
are Copyright 2012 Twitter, Inc and licensed under the Apache
License v2.0.
Modifications have been made by us to create a project specific look & feel.

The rest of the code is Copyright 2012 by Mats Strandberg and Jan Grape and
licensed under the Apache License v2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.


Setting Up Development Environment
----------------------------------

### Basics ###

In order to set up the development environment you need to:

 - Clone this project from GitHub
 - Install Play 2.0
 - Install Heroku Toolkit - https://toolbelt.heroku.com/ (Optional, only needed to deploy to Heroku)
 - ```play eclipsify``` or ```play idea```

### "Persistent Database" - Postgres ###

When running SignUp as described above, the H2 database is used.
The H2 database is running in-memory, in-process. This is handy,
because of the simple setup, but each time you stop Play your database
will be gone. It would be nice to have a persistent database,
e.g. a disk based database. When running on Heroku, Postgres will be used.
For these reasons, a local installation of Postgres is good.

To install Postgres on MacOS X:

 - Download Postgres from [postgres.org](http://www.postgresql.org/)
 - Install a standard Postgres (You may have to restart your Mac as the installation fiddles with shared memory)
 - Set the password for the DATABASE (super)user postgres when prompted
 - Finish the installation
 - On MacOS X: A UNIX user is created: 'PostgreSQL'
   This user seem to get some password that you cannot find out.
   Reset the password of the UNIX user 'PostgreSQL' to something you know, and something safe, as this is a real MacOS X user.
 - Do some stuff from the command prompt to create a database and a database user

```
$ su - PostgreSQL
Password: <Enter the password of your UNIX user 'PostgreSQL'>
$ createdb signup
Password: <Enter the password of your DATABASE user 'postgres'>
$ psql -s signup
Password: <Enter the password of your DATABASE user 'postgres'>
signup=# create user signup4 password 's7p2+';
signup=# grant all privileges on database signup to signup4;
```

Now you can start SignUp with Postgres using

./playrunpostgres.sh

Run SignUp
----------

Once you have the development environment set up you should be able to do

```play run```

And the direct your browser to
[http://localhost:9000](http://localhost:9000)

If you have a local Postgres (installed as described above) you may start SignUp using

./playrunpostgres.sh

Deploy
------

Deployment is done to Heroku.

It runs at [http://signup4.herokuapp.com](http://signup4.herokuapp.com)

To be able to deploy to Heroku you must:

* Install Heroku toolbelt.
* Do a 'heroku login'.
* cd into the root of this application
* git remote add staging git@heroku.com:signup4.git
* git push staging master

I (Mats) had to do a 'heroku keys:add ~/.ssh/id_rsa.pub' on one of my
machines as I first got 'Permission denied (publickey).' when trying
to 'git push heroku master'



newrelic/CHANGELOG
CHANGELOG
=========

v2.21.3
-------

* Fix: Agent did not run with J2SE 5.0 (aka JDK 1.5)

  Release 2.21.0 - 2.21.2 did not run correctly on J2SE 5.0. On startup,
  the agent reports:

    Unable to start New Relic agent: java.lang.UnsupportedClassVersionError:
    Bad version number in .class file 

  If you are running J2SE 5.0, please update to agent version 2.21.3.


v2.21.2
-------

* Fix: Longer application startup time

  Release 2.21.0 introduced an issue that could cause application startup
  to be slower than in previously releases. This fix returns application
  startup time to normal.

* Fix: File location changes

  Release 2.21.0 introduced a change to file locations within the
  newrelic zip file. This fix reverts the locations to what they were
  previously.


v2.21.1
-------

* Fix: In 2.21.0, calls to NewRelic class can cause NoClassDefFoundError

  In 2.21.0, if your application calls the NewRelic class and does
  not include the newrelic-api.jar, the application would throw a
  NoClassDefFoundError. With this fix, the separate newrelic-api.jar is
  not required. However, it is recommended that your application include
  the newrelic-api.jar so that calls to the NewRelic API resolve with or
  without the agent installed.


v2.21.0
-------

* X-Ray sessions

  Adds support for X-Ray sessions.  An X-Ray session collects
  Transaction Traces and a thread profile for a Key Transaction.

* Fix: In 2.20.0 an @Trace annotation ignored the transaction for Apdex 

  An @Trace annotation on a method called in a web transaction had
  the side-effect of ignoring the transaction for calculating the Apdex score.

* Fix: HTTP status code for Errors 
    
  If the HTTP status code for a web transaction was less than 400, the HTTP
  status code reported in the error was 500.  The actual status code for the
  web request is now reported.
    
* Fix: Spring exception handling
    
  The Exception argument in the Spring DispatcherServlet processHandlerException method
  was reported as an Error with an HTTP status code of 500.  The actual status code
  in the HTTP response is now reported.
    
* Fix: Reading Post Parameters for Resin 3

  Prior to this fix, post parameters in Resin 3 were always read using the 
  default properties even if the application tried to read the post parameters 
  using non-default properties or by just grabbing the request input stream.

* Fix: JVM Tab Metrics Not Showing up for Tomcat Applications

  Prior to this fix, if the service name in service.xml was not Catalina, then
  metrics would not appear under the JVM tab. Additionally, if you are using an 
  embedded Tomcat 5/6, metrics should now appear under the JVM tab.

* Fix: Improved compatibility with Java 1.5

  In the custom xml validator, there were references to String.isEmpty() which
  was not available until Java 1.6.  Those have been removed.

* Fix: Reducing DNS problems by not automatically adding '.' to the domain name

  By default the '.' is included. But if changed, then the '.' won't be added.

* Fix: JVM tab metrics showing for all Applications In JVM

  Metrics under the Thread, Session, and Transaction tabs used to only appear
  under the main application. The data will now be shown for all applications
  running in the specific JVM.

* Fix: Using a Proxy now works with SSL Enabled 

  In prior releases proxy authentication would fail when SSL was enabled.
  We now do proxy authentication using the java.net.Authenticator.
  
* Fix: Akka error reporting

  If an actor throws an exception or the message times out, an error is reported.
  Previously, no error was reported.
  

v2.20.0
-------
* Glassfish 2.1 JVM Tab

  The JVM tab will now show thread, session, and transaction metrics for 
  Glassfish 2.1.
  
* Fix: Solr Cache Metrics 

  The solr cache metrics do not appear with versions 2.18 and 2.19
  of the agent. This has been fixed and all solr metrics should now appear.
  
* Fix: NullPointerException in New Relic Logs When NewRelic API called 

  In certain circumstances, a call to a method on the NewRelic class can
  generate a NullPointerException in the agent logs. It does not cause a
  NullPointerException in your application.
  
* Fix: Support Custom Extension XML File for WebSphere with JRE

  Prior to this fix, an attempt to use a custom extension XML file with a IBM JRE
  would lead to a ClassNotFoundException.

v2.19.1
-------

* Fix for 2.19.0 known issue:  Play now captures GET and POST
  parameters correctly if `capture_params` is enabled.

* Disabled overly ambitious instrumentation used to time template rendering for Play 2.
  Resulted in unintended classes being instrumented.

v2.19.0
-------

* Play 2 Instrumentation

  Provides insight into Play 2.x Java and Scala applications. Adds
  instrumentation to Netty server, Akka actor calls, Scala Promises and Futures,
  and the WS API calls.

  Known issue: Request parameters are not captured properly
  even if `capture_params` is enabled.

* Support added for Jetty 9 (tested on 9.0.0 through Jetty 9.0.3).

* Fix: Agent can cause java.lang.VerifyError on code compiled with Java 7

  In certain cases, if your application code or third party libraries are
  compiled with Java 7, you may get a java.lang.VerifyError. This fix allows
  the agent to work correctly in those cases.

* Fix: The Java Agent detects status codes for Apache Felix correctly by default.
  The property take_last_status is now deprecated.

v2.18.0
-------
* New metrics are displayed under the JVM tab

  While the metrics displayed differ between application servers, common
  metrics include loaded class count, active thread count, active thread count
  per thread pool, active session count per application, and active
  transaction count. In order to see these metrics, JMX must be enabled on
  your app server.

* Fix: When naming transactions, the property `enable_auto_transaction_naming`
  should be honored


project/Build.scala
import sbt._
import Keys._
import play.Project._

object ApplicationBuild extends Build {

  val appName = "signup"
  val appVersion = "1.0-SNAPSHOT"

  val appDependencies = Seq(
    "postgresql" % "postgresql" % "9.1-901-1.jdbc4",
    "jp.t2v" %% "play2.auth" % "0.9",
    "jp.t2v" %% "play2.auth.test" % "0.9" % "test",
    "com.typesafe" %% "play-plugins-mailer" % "2.1.0",
    "com.newrelic.agent.java" % "newrelic-api" % "2.19.+",
    jdbc,
    anorm
  )

  // Only compile the LESS files listed here. Others will be included by the top ones.
  def customLessEntryPoints(base: File): PathFinder = (
    (base / "app" / "assets" / "stylesheets" / "bootstrap" * "bootstrap.less") +++
    (base / "app" / "assets" / "stylesheets" / "bootstrap" * "responsive.less")
  )

  val main = play.Project(appName, appVersion, appDependencies).settings(
    resolvers += "t2v.jp repo" at "http://www.t2v.jp/maven-repo/",
    lessEntryPoints <<= baseDirectory(customLessEntryPoints),
    scalacOptions += "-feature"
  )
}
